[bug-fix] Set the db sync height

rajarshimaitra opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

Fixes #719 

Previously we weren't setting the db sync height in populate_test_db
macro even when current height is provided.. This creates a bug that
get_funded_wallet will return 0 balance.

This PR fixes the populate_test_db macro and updates tests which were
previously dependent on the unsynced wallet behavior.

### Notes to the reviewers

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -2862,7 +2898,21 @@ pub(crate) mod test {
+                @tx ( (@external descriptors, 0) => 50_000 ) (@confirmations 1)
+        };
+
+        // Add the transaction to our db, but do not sync the db.
```
Same here, a comment would be really helpful!

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -2364,7 +2385,22 @@ pub(crate) mod test {
+                @tx ( (@external descriptors, 0) => 50_000 ) (@confirmations 1)
+        };
+
+        // Add the transaction to our db, but do not sync the db.
```
Can we add a comment here explaining why we don't want to sync the db?

rajarshimaitra replied:
Done.. Yes would be helpful..

rajarshimaitra commented:
Updated with review comments..

vladimirfomene commented on src/wallet/mod.rs:
```diff
@@ -2364,7 +2385,23 @@ pub(crate) mod test {
+        };
+
+        // Add the transaction to our db, but do not sync the db. Unsynced db
+        // should trigger the default sequqence value for a new transaction as 0xFFFFFFFF
```
Nits: spelling error on `sequence`

rajarshimaitra replied:
Thanks.. Done..

vladimirfomene commented on src/wallet/mod.rs:
```diff
@@ -2862,7 +2899,22 @@ pub(crate) mod test {
+        };
+
+        // Add the transaction to our db, but do not sync the db. Unsynced db
+        // should trigger the default sequqence value for a new transaction as 0xFFFFFFFF
```
Nits: spelling error on `sequence`

vladimirfomene reviewed (commented):
utACK

rajarshimaitra commented:
Thanks @vladimirfomene for the catches.. Updated..

danielabrozzoni reviewed (approved):
utACK b8ce842ce01752ce693759bbfa5007f5619c2424

afilini commented on src/database/memory.rs:
```diff
@@ -512,13 +513,29 @@ macro_rules! populate_test_db {
+            })
+        } else {
+            None
+        };
```
I don't think it's ok to silently drop the `min_confirmations` if the current height is not provided. I would rather keep what we had before with a better error message, by changing the `unwrap()` into an `expect()`.

rajarshimaitra replied:
Makes sense..

afilini commented on src/database/memory.rs:
```diff
@@ -512,13 +513,29 @@ macro_rules! populate_test_db {
+                    timestamp: 0
+                }
+            };
+            db.set_sync_time(sync_time).unwrap();
```
I think here you should only keep the most recent sync time, so instead of just setting it you should first check if something is present and compare the two

rajarshimaitra replied:
It should.. Done.,.

rajarshimaitra committed:
Set the db sync height

Previously we weren't setting the db sync height in populate_test_db
macro even when current height is provided.. This creates a bug that
get_funded_wallet will return 0 balance.

This PR fixes the populate_test_db macro and updates tests which were
previously dependent on the unsynced wallet behavior.

rajarshimaitra commented:
Thanks @afilini  for the look.. Updated as suggested.. In the last dev call we talked about releasing this as a patch version for some downstream libs to update.. See https://github.com/bitcoindevkit/bdk-cli/pull/102#issuecomment-1226928903

danielabrozzoni commented:
Sorry I didn't notice the two issues that Alekos has found, I should have inspected the code better.

re-utACK a72a97b663fa71cb66e91f1187dc257762dbb2a7

notmandatory commented:
@rajarshimaitra  even though we talked about doing a fix release for this, since this is a test framework fix and code freeze for 0.22 is next Wed. (Aug 31), can we skip doing a fix release only for this PR and wait for 0.22 to put it out (Sept 7)?

rajarshimaitra commented:
@notmandatory sounds good to me.. Only needed for the bdk-cli PR.. That can wait a bit..

afilini commented on src/database/memory.rs:
```diff
@@ -512,14 +513,32 @@ macro_rules! populate_test_db {
+        // If any of them is None, the other will be applied instead.
+        // If both are None, this will not be set.
+        if let Some(height) = db.get_sync_time().unwrap()
+                                .and_then(|sync_time| Some(sync_time.block_time.height))
```
nit: You can replace `and_then` with `map` because you always return `Some()`

rajarshimaitra commented:
Updated and rebased..

afilini reviewed (commented):
ACK 08668ac46247d527cc53af5b6f359b1fa4e3b6aa
